README.md
[![Build Status](https://travis-ci.org/mcmillhj/Net-Easypost.svg?branch=master)](https://travis-ci.org/mcmillhj/Net-Easypost)
[![Coverage Status](https://coveralls.io/repos/mcmillhj/Net-Easypost/badge.png?branch=master)](https://coveralls.io/r/mcmillhj/Net-Easypost?branch=master)
[![Kwalitee status](http://cpants.cpanauthors.org/dist/Net-Easypost.png)](http://cpants.charsbar.org/dist/overview/Net-Easypost)

# NAME

Net::Easypost - Perl client for the Easypost web service

# VERSION

version 0.18

# SYNOPSIS

    use Net::Easypost;

    my $to = Net::Easypost::Address->new(
       name    => 'Johnathan Smith',
       street1 => '710 East Water Street',
       city    => 'Charlottesville',
       state   => 'VA',
       zip     => '22902',
       phone   => '(434)555-5555',
       country => 'US',
    );

    my $from = Net::Easypost::Address->new(
       name    => 'Jarrett Streebin',
       phone   => '3237078576',
       city    => 'Half Moon Bay',
       street1 => '310 Granelli Ave',
       state   => 'CA',
       zip     => '94019',
       country => 'US',
    );

    my $parcel = Net::Easypost::Parcel->new(
       length => 10.0,
       width  => 5.0,
       height => 8.0,
       weight => 10.0,
    );

    my $shipment = Net::Easypost::Shipment->new(
       to_address   => $to,
       from_address => $from,
       parcel       => $parcel,
    );

    my $ezpost = Net::Easypost->new;
    my $label = $ezpost->buy_label($shipment, ('rate' => 'lowest'));

    printf("You paid \$%0.2f for your label to %s\n", $label->rate->rate, $to);
    $label->save;
    printf "Your postage label has been saved to '" . $label->filename . "'\n";

# OVERVIEW

This is a Perl client for the postage API at [Easypost](https://www.easypost.com/docs/api). Consider this
API at beta quality mostly because some of these library calls have an inconsistent input
parameter interface which I'm not super happy about. Still, there's enough here to get
meaningful work done, and any future changes will be fairly cosmetic.

Please note! **All API errors are fatal via croak**. If you need to catch errors more gracefully, I
recommend using [Try::Tiny](https://metacpan.org/pod/Try::Tiny) in your implementation.

API key:

You must have your API key stored in an environment variable named 
EASYPOST\_API\_KEY

# ATTRIBUTES

## requester

HTTP client to POST and GET

# METHODS

## verify\_address

This method attempts to validate an address. This call expects to take the same parameters
(in a hashref) or an instance of [Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address), namely:

- street1
- street2
- city
- state
- zip
- country

You may omit some of these attributes like city, state if you supply a zip, or
zip if you supply a city, state.

This call returns a new [Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address) object.

Along with the validated address, the `phone` and `name` fields will be
copied from the input parameters, if they're set.

The verification works only for addresses in US. If you pass a country
other than US (the default), a warning will be issued, but the
[Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address) object will be returned.

## get\_rates

This method will get postage rates between two zip codes. It takes the following input parameters:

- to => an instance of [Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address)
- from => an instance of [Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address)
- parcel => an instance of [Net::Easypost::Parcel](https://metacpan.org/pod/Net::Easypost::Parcel)

This call returns an array of [Net::Easypost::Rate](https://metacpan.org/pod/Net::Easypost::Rate) objects in an arbitrary order.

## buy\_label

This method will attempt to purchase postage and generate a shipping label.

It takes as input:

- A [Net::Easypost::Shipment](https://metacpan.org/pod/Net::Easypost::Shipment) object
- A [Net::Easypost::Rate](https://metacpan.org/pod/Net::Easypost::Rate) object

It returns a [Net::Easypost::Label](https://metacpan.org/pod/Net::Easypost::Label) object.

## get\_label

This method retrieves a label from a past purchase. It takes the label filename as its
only input parameter. It returns a [Net::Easypost::Label](https://metacpan.org/pod/Net::Easypost::Label) object.

## list\_labels

This method returns an arrayref with all past purchased label filenames. It takes no
input parameters.

# SUPPORT

Please report any bugs or feature requests to "bug-net-easypost at
rt.cpan.org", or through the web interface at
[http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-Easypost](http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-Easypost).  I will
be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

Or, if you wish, you may report bugs/features on Github's Issue Tracker.
[https://github.com/mrallen1/Net-Easypost/issues](https://github.com/mrallen1/Net-Easypost/issues)

# SEE ALSO

- [Easypost API docs](https://www.easypost.com/docs/api)

# AUTHOR

Mark Allen <mrallen1@yahoo.com>, Hunter McMillen <mcmillhj@gmail.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Change history for Net-Easypost

{{$NEXT}}

0.18        2016-01-05
            - Add JSON as dependency in dist.ini
0.17        2016-01-05
            - Remove Mojo::UserAgent in favor of HTTP::Tiny
            
0.16        2015-12-29
            - Minor changes, mostly formatting and build configuration
            - Now have an explicit cpanfile and Makefile.PL in the release in case some don't want to use Dist::Zilla
0.15        2015-03-18

0.14        2015-03-04

0.13        2014-05-27

0.12        2014-05-22

0.11        2014-05-21

0.09        2013-07-31
            Don't be so clever when passed a hashref in get_rates
            (Fixes RT#87493)
            Fix failing tests in t/01-basic.t
            Add test for RT#87493
            Add CPANfile to dist.ini

0.08        2013-05-19
            Seriously. Fix rate number test.

0.07        2013-05-16
            Check carriers against a regex
            For reals, make IO::Socket::SSL 
                a dependency

0.06        2013-05-15
            Make IO::Socket::SSL a prereq
            Check rate against regex

0.05        2013-05-14
            Fix post request method due to 
                Mojo deprecation warnings

0.04        2013-05-13
            Changed abstract
            Fixed dependency (Thanks Graham Knop)
            Fixed t/01-basic.t

0.03        2013-01-19
            Update URLs in POD
            Removed Data::Printer
            Return actual endpoint error codes 
            instead of inspecting JSON.

0.02        2013-01-13
            New endpoint
            Updated test suite

0.01        2012-12-04
            Initial release


lib/Net/Easypost.pm
package Net::Easypost;

use Data::Dumper;
use Carp qw(croak);
use Hash::Merge::Simple qw(merge);
use Moo;

use Net::Easypost::Address;
use Net::Easypost::Label;
use Net::Easypost::Parcel;
use Net::Easypost::Rate;
use Net::Easypost::Request;
use Net::Easypost::Shipment;

# ABSTRACT: Perl client for the Easypost web service

=head1 SYNOPSIS

   use Net::Easypost;

   my $to = Net::Easypost::Address->new(
      name    => 'Johnathan Smith',
      street1 => '710 East Water Street',
      city    => 'Charlottesville',
      state   => 'VA',
      zip     => '22902',
      phone   => '(434)555-5555',
      country => 'US',
   );

   my $from = Net::Easypost::Address->new(
      name    => 'Jarrett Streebin',
      phone   => '3237078576',
      city    => 'Half Moon Bay',
      street1 => '310 Granelli Ave',
      state   => 'CA',
      zip     => '94019',
      country => 'US',
   );

   my $parcel = Net::Easypost::Parcel->new(
      length => 10.0,
      width  => 5.0,
      height => 8.0,
      weight => 10.0,
   );

   my $shipment = Net::Easypost::Shipment->new(
      to_address   => $to,
      from_address => $from,
      parcel       => $parcel,
   );

   my $ezpost = Net::Easypost->new;
   my $label = $ezpost->buy_label($shipment, ('rate' => 'lowest'));

   printf("You paid \$%0.2f for your label to %s\n", $label->rate->rate, $to);
   $label->save;
   printf "Your postage label has been saved to '" . $label->filename . "'\n";

=head1 OVERVIEW

This is a Perl client for the postage API at L<Easypost|https://www.easypost.com/docs/api>. Consider this
API at beta quality mostly because some of these library calls have an inconsistent input
parameter interface which I'm not super happy about. Still, there's enough here to get
meaningful work done, and any future changes will be fairly cosmetic.

Please note! B<All API errors are fatal via croak>. If you need to catch errors more gracefully, I
recommend using L<Try::Tiny> in your implementation.

API key:

You must have your API key stored in an environment variable named 
EASYPOST_API_KEY

=cut

=attr requester

HTTP client to POST and GET

=cut

has 'requester' => (
    is      => 'ro',
    lazy    => 1,
    default => sub { return Net::Easypost::Request->new }
);

=method verify_address

This method attempts to validate an address. This call expects to take the same parameters
(in a hashref) or an instance of L<Net::Easypost::Address>, namely:

=over

=item * street1

=item * street2

=item * city

=item * state

=item * zip

=item * country

=back

You may omit some of these attributes like city, state if you supply a zip, or
zip if you supply a city, state.

This call returns a new L<Net::Easypost::Address> object.

Along with the validated address, the C<phone> and C<name> fields will be
copied from the input parameters, if they're set.

The verification works only for addresses in US. If you pass a country
other than US (the default), a warning will be issued, but the
L<Net::Easypost::Address> object will be returned.


=cut

sub verify_address {
    my ($self, $params) = @_;

    if ( ref($params) eq 'HASH' ) {
        return Net::Easypost::Address->new( $params )->verify;
    }
    elsif ( ref($params) eq 'Net::Easypost::Address' ) {
        return $params->verify;
    }
    else {
        croak "verify_address expects either a hashref or an instance of Net::Easypost::Address\n";
    }
}

=method get_rates

This method will get postage rates between two zip codes. It takes the following input parameters:

=over

=item * to => an instance of L<Net::Easypost::Address>

=item * from => an instance of L<Net::Easypost::Address>

=item * parcel => an instance of L<Net::Easypost::Parcel>

=back

This call returns an array of L<Net::Easypost::Rate> objects in an arbitrary order.

=cut

sub get_rates {
    my $self = shift; # shift is important here

    my $params;
    if ( scalar @_ == 1 ) {
        if ( ref( $_[0] ) ne 'HASH' ) {
            croak 'get_rates expects a hashref not a '. ref($params) .'\n';
        }
        else {
            $params = shift;
        }
    }
    else {
        $params = { @_ };
    }

    return Net::Easypost::Shipment->new(
        to_address   => $params->{to},
        from_address => $params->{from},
        parcel       => $params->{parcel},
    )->rates;
}

=method buy_label

This method will attempt to purchase postage and generate a shipping label.

It takes as input:

=over

=item * A L<Net::Easypost::Shipment> object

=item * A L<Net::Easypost::Rate> object

=back

It returns a L<Net::Easypost::Label> object.

=cut

sub buy_label {
    my ($self, $shipment, %options) = @_;

    croak 'Buy label expects a parameter of type Net::Easypost::Shipment'
        unless $shipment || ref($shipment) ne 'Net::Easypost::Shipment';

    return $shipment->buy(%options);
}

=method get_label

This method retrieves a label from a past purchase. It takes the label filename as its
only input parameter. It returns a L<Net::Easypost::Label> object.

=cut

sub get_label {
    my ($self, $label_filename) = @_;

    my $resp = $self->requester->post(
        '/postage/get', 
        { label_file_name => $label_filename } 
    );

    return Net::Easypost::Label->new(
        rate          => Net::Easypost::Rate->new( $resp->{rate} ),
        tracking_code => $resp->{tracking_code},
        filename      => $resp->{label_file_name},
        filetype      => $resp->{label_file_type},
        url           => $resp->{label_url}
    );
}

=method list_labels

This method returns an arrayref with all past purchased label filenames. It takes no
input parameters.

=cut

sub list_labels {
    my ($self) = @_;

    my $resp = $self->requester->get( 
        $self->requester->_build_url('/postage/list') 
    );

    return $resp->{postages};
}

=head1 SUPPORT

Please report any bugs or feature requests to "bug-net-easypost at
rt.cpan.org", or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-Easypost>.  I will
be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

Or, if you wish, you may report bugs/features on Github's Issue Tracker.
L<https://github.com/mrallen1/Net-Easypost/issues>

=head1 SEE ALSO

=over

=item * L<Easypost API docs|https://www.easypost.com/docs/api>

=back

=cut

1;
